airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeS3StagingSqlOperations.java

```diff
@@ -0,0 +1,207 @@
+            csvPrinter.printRecord(csvSheetGenerator.getDataRow(id, recordMessage));
+          }
+        }
+      }
+    } catch (final Exception e) {
+      LOGGER.error("Failed to load data into stage {}", stage, e);
+      hasFailed = true;
+      throw new RuntimeException(e);
+    } finally {
+      if (hasFailed) {
+        uploadManager.abort();
+      } else {
+        uploadManager.complete();
+      }
+    }
+    if (!s3Client.doesObjectExist(bucket, objectKey)) {
+      LOGGER.error("Failed to upload data into stage, object {} not found", objectKey);
+      throw new RuntimeException("Upload failed");
+    }
+  }
+
+  @Override
+  public void createStageIfNotExists(final JdbcDatabase database, final String stageName) {
+    final String bucket = s3Config.getBucketName();
+    if (!s3Client.doesBucketExistV2(bucket)) {
+      LOGGER.info("Bucket {} does not exist; creating...", bucket);
+      s3Client.createBucket(bucket);
+      LOGGER.info("Bucket {} has been created.", bucket);
+    }
+    if (!s3Client.doesObjectExist(bucket, stageName)) {
+      LOGGER.info("Stage {}/{} does not exist; creating...", bucket, stageName);
+      s3Client.putObject(bucket, stageName.endsWith("/") ? stageName : stageName + "/", "");
+      LOGGER.info("Stage {}/{} has been created.", bucket, stageName);
+    }
+  }
+
+  @Override
+  public void copyIntoTmpTableFromStage(final JdbcDatabase database, final String path, final String dstTableName, final String schemaName) {
+    LOGGER.info("Starting copy to tmp table from stage: {} in destination from stage: {}, schema: {}, .", dstTableName, path, schemaName);
+    final var copyQuery = String.format(
+        "COPY INTO %s.%s FROM '%s' "
+            + "CREDENTIALS=(aws_key_id='%s' aws_secret_key='%s') "
+            + "file_format = (type = csv field_delimiter = ',' skip_header = 0 FIELD_OPTIONALLY_ENCLOSED_BY = '\"');",
+        schemaName,
+        dstTableName,
+        generateBucketPath(path),
+        s3Config.getAccessKeyId(),
+        s3Config.getSecretAccessKey());
+    Exceptions.toRuntime(() -> database.execute(copyQuery));
+    LOGGER.info("Copy to tmp table {}.{} in destination complete.", schemaName, dstTableName);
+  }
+
+  private String generateBucketPath(final String stage) {
+    return "s3://" + s3Config.getBucketName() + "/" + stage;
+  }
+
+  @Override
+  public void dropStageIfExists(final JdbcDatabase database, final String stageName) {
+    LOGGER.info("Dropping stage {}...", stageName);
+    final String bucket = s3Config.getBucketName();
+    if (s3Client.doesObjectExist(bucket, stageName)) {
+      s3Client.deleteObject(bucket, stageName);
+    }
+    LOGGER.info("Stage object {} has been deleted...", stageName);
+  }
+
+  @Override
+  public void cleanUpStage(final JdbcDatabase database, final String stageName) {
+    final String bucket = s3Config.getBucketName();
+    ObjectListing objects = s3Client.listObjects(bucket, stageName);
+    while (objects.isTruncated()) {
```

Comment:
do we not need to delete anything if the list is not truncated?

Reply:
correct! thanks